content/post/2020-03-05-user-preferences/index.Rmd

```diff
@@ -0,0 +1,254 @@
+rappdirs::user_cache_dir("rhub")
+```
+
+On top of these non-R specific standard locations, we'll also mention the standard homes of R _options_ and _environment variables_, .Rprofile and .Renviron.
+
+## User preferences
+
+As written in [Android developer guidance](https://developer.android.com/training/id-auth/identify) and probably every customer service guide ever, _"Everyone likes it when you remember their name"_. 
+Everyone probably likes it too when the barista at their favourite coffee shop remembers their usual order. 
+As an R package developer, what can you do for your R package to correctly assess user preferences and settings?
+
+### Using options
+
+In R, `options` _allow the user to set and examine a variety of global options which affect the way in which R computes and displays its results_. For instance, for the usethis package, the [`usethis.quiet` option can control whether usethis is chatty](https://github.com/r-lib/usethis/blob/c0acd1b5e43f03773a5934c1d937bf70b77a2557/principles.md#communicating-with-the-user)[^1]. Users either:
+
+* write `options(usethis.quiet = TRUE)` at the beginning of a script or directly in the console;
+
+* or write that same line in their [.Rprofile that's loaded before every session](https://rstats.wtf/r-startup.html), which is more persistent. 
+
+Users can use a project-level or more global user-level .Rprofile. 
+The use of a project-level .Rprofile overrides the user-level .Rprofile unless the project-level .Rprofile contains the following lines [as mentioned in the `blogdown` book](https://bookdown.org/yihui/blogdown/global-options.html):
+
+```r
+# in .Rprofile of the project
+if (file.exists('~/.Rprofile')) {
+  base::sys.source('~/.Rprofile', envir = environment())
+}
+# then set project options
+```
+
+For more startup tweaks, the user could adopt [the `startup` package](https://cran.r-project.org/web/packages/startup/index.html).
+
+As a package developer in your code you can retrieve options by using `getOption()` whose second argument is a fallback for when the option hasn't been set by the user.
+Note that an option can be any R object.
+
+```{r options}
+options(blabla.foo = TRUE)
+if (getOption("blabla.foo", FALSE)) {
+  message("foo!")
+}
+
+options(blabla.bar = mean)
+getOption("blabla.bar")(c(1:7))
+```
+
+The use of options in the .Rprofile startup file is great for workflow packages like `usethis`, `blogdown`, etc., but shouldn't be used for, say, arguments influencing the results of a statistical function.
+
+### Using environment variables
+
+Environment variables, found via `Sys.getenv()` rather than `getOption()`, are appropriate for storing secrets (like [`GITHUB_PAT` for the `gh` package](https://github.com/r-lib/gh#tokens)) or the path to secrets on disk (like [`TWITTER_PAT` for `rtweet`](https://rtweet.info/articles/auth.html)), or not secrets (e.g. [the browser to use for `chromote`](https://github.com/rstudio/chromote#specifying-which-browser-to-use)).
```

Comment:
In theory options are just as appropriate for secrets. But env vars work for other software, e.g. setting `GITHUB_TOKEN` will also work in node.js.

Reply:
but not if you set it in .Renviron? I.e. node.js won't look into .Renviron?

Reply:
Yeah, but you can set it in the shell, or as a user-level env var in Windows.

Reply:
Anyway, this is not terribly important. But I think the statement that env vars are appropriate for storing secrets does not (always) hold, from a security point of view. We just couldn't really do any better.

Reply:
We can say that many people use env vars for secrets, that's certainly true.

Reply:
good point, I'll write "often used".

Reply:
(and mention credential stores are more secure)